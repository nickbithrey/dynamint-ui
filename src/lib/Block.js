import React from 'react';
import { Stack } from 'office-ui-fabric-react/lib/Stack';

const Block = ({children, ...remainingProps}) => (
	<Stack.Item {...remainingProps}>
		{children}
	</Stack.Item>
);

export default Block;