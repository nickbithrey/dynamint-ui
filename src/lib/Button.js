import React from 'react';
import PropTypes from 'prop-types';
import { 
	PrimaryButton, 
	DefaultButton, 
	CommandBarButton, 
	IconButton 
} from 'office-ui-fabric-react/lib/Button';

const buttonTypes = {
	standard: DefaultButton,
	primary: PrimaryButton,
	command: CommandBarButton,
	icon: IconButton
};

const Button = ({text, onClick, type}) => {
	const Btn = buttonTypes[type];
	return <Btn text={text} onClick={onClick} />
};

Button.defaultProps = {
	type: 'standard'
};

Button.propTypes = {
	text: PropTypes.string.isRequired,
	onClick: PropTypes.func
}

export default Button;

export const newBtn = (text, onClick, condition = () => true) => (
	{
		text,
		onClick,
		condition,
		type: 'standard'
	}
);

export const newPrimaryBtn = (text, onClick, condition = () => true) => (
	{
		text,
		onClick,
		condition,
		type: 'primary'
	}
);