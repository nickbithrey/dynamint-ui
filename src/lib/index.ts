export interface Key {
    key: string | number;
}

export interface Condition<T> {
    condition: (input: T) => boolean;
}

export interface Loading {
    loading: boolean;
}

export interface Items<T> {
    items: Array<T>;
}

export interface URI {
    uri: string;
}

export interface Display {
    readOnly: boolean;
}

export interface IURILink {
    href: string;
}