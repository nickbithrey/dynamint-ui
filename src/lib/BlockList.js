import React from 'react';
import PropTypes from 'prop-types';
import { List } from 'office-ui-fabric-react/lib/List';

const renderCell = (tag, updateEntry) => {
	const CellTag = tag;
	return (item, index) => {
		return <CellTag item={item} index={index} update={updateEntry} />;
	};
};

const BlockList = ({items, tag, updateEntry}) => {
	return (
		<List
	        items={items}
	        onRenderCell={renderCell(tag, updateEntry)}
	    />
	);
};

BlockList.propTypes = {
	items: PropTypes.arrayOf(PropTypes.object).isRequired,
	tag: PropTypes.func.isRequired
}

export default BlockList;