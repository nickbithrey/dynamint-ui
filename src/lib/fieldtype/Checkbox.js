import React from 'react';
import PropTypes from 'prop-types';
import { TextField } from 'office-ui-fabric-react/lib/TextField';
import { Checkbox as CheckboxField } from 'office-ui-fabric-react/lib/Checkbox';

const Checkbox = ({name, label, value, update, readOnly, ...remainingProps}) => (
	<CheckboxField 
		label={label} 
		checked={value} 
		onChange={update} 
		{...remainingProps}
	/>
);

Checkbox.defaultProps = {
	value: false
}

Checkbox.propTypes = {
	name: PropTypes.string.isRequired,
	label: PropTypes.string,
	value: PropTypes.bool,
	update: PropTypes.func
};

export default Checkbox;