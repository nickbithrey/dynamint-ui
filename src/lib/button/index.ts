import { Button, IButton } from './Button';
import { ButtonContainer, IButtonContainer } from './ButtonContainer';

export { 
    Button,
    IButton,
    ButtonContainer,
    IButtonContainer
};