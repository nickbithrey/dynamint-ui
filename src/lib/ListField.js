import React from 'react';
import PropTypes from 'prop-types';
import Field from './Field';
import Button from '~/lib/Button';
import { Rowise, Block } from '~/lib/Grid';
import { DetailsList, DetailsRow, DetailsListLayoutMode, SelectionMode, Selection } from 'office-ui-fabric-react/lib/DetailsList';
import { Checkbox } from 'office-ui-fabric-react/lib/Checkbox';

export default class ListField extends React.Component {
	
	constructor(props) {
		super(props);
		this.update = this.update.bind(this);
		this.addElement = this.addElement.bind(this);
		
		this.selection = new Selection({
		    onSelectionChanged: () => {
		        this.setState({
		            selectionDetails: this.getSelectionDetails()
		        });
		    }
	    });
		this.state = {};
	}
	
	getSelectionDetails() {
		if (this.selection.getSelectedCount() > 0) {
			this.setState({
				index: this.selection.getSelectedCount()
			});
			return this.selection;
		}
	}
	
	update(element, index) {
		const list = [...this.props.value];
		list[index] = element;
		this.props.update({
			name: this.props.name,
			value: list
		});
		this.selection.toggleIndexSelected(index);
	}
	
	addElement(e) {
		e.preventDefault();
		const list = [...this.props.value];
		list.push({...this.props.newElement});
		this.props.update({
			name: this.props.name,
			value: list
		});
//		this.selection.toggleIndexSelected(list.length);
	}
	
	render() {
		const {
			value,
			columns,
			...remainingProps
		} = this.props;
		return (
			<div>
				<Rowise tokens={{childrenGap: 5}}>
					<Block>
						<DetailsList
							items={[...value]}
						    compact={false}
						    columns={columns}
						    selectionMode={SelectionMode.single}
						    setKey="set"
						    layoutMode={DetailsListLayoutMode.justified}
						    isHeaderVisible={true}
							onRenderItemColumn={this.renderColumn}
							onRenderRow={this.renderRow(this.update, this.props.tag)}
						    selectionPreservedOnEmptyClick={true}
							selection={this.selection}
						    enterModalSelectionOnTouch={true}
						    ariaLabelForSelectionColumn="Toggle selection"
						    ariaLabelForSelectAllCheckbox="Toggle selection for all items"
					    />
					</Block>
					<Block>
						<Button text="Add" onClick={this.addElement} />
					</Block>
				</Rowise>
			</div>
		)
	}
	
	renderColumn(item, index, column) {
		  const fieldContent = item[column.fieldName];

		  switch (column.type) {
		    case 'boolean':
		      return <span><Checkbox checked={fieldContent} /></span>
		    default:
		      return <span>{fieldContent}</span>;
		  }
		}
	
	renderRow(fieldUpdate, tag) {
		return function(props) {
			if (props.selection.getSelectedIndices().includes(props.itemIndex)) {
				const Tag = listUpdate(tag);
				return (
					<div>
						<DetailsRow {...props} />
						<Tag fieldUpdate={fieldUpdate} {...props} />
					</div>
				)
			}
			return <DetailsRow {...props} />
		}
	}
	
}

export function listUpdate(WrappedComponent) {
	return class ListUpdate extends React.Component {
		constructor(props) {
			super(props);
			this.update = this.update.bind(this);
			this.updateField = this.updateField.bind(this);
			this.index = props.itemIndex;
			this.state = {...props.item};
		}
		
		update(e) {
			e.preventDefault();
			this.props.fieldUpdate(this.state, this.index);
		}
		
		updateField({name, value}) {
			this.setState({
				[name]: value
			});
		}
		
		render() {
			return (
				<div>
					<Rowise>
						<Block>
							<WrappedComponent updateField={this.updateField} {...this.state} />
						</Block>
						<Block>
							<Button text="Submit" onClick={this.update} />
						</Block>
					</Rowise>
				</div>
			)
		}
	}
}

ListField.defaultProps = {
	readOnly: true,
	value: []
};

ListField.propTypes = {
	name: PropTypes.string.isRequired,
	label: PropTypes.string,
	value: PropTypes.arrayOf(PropTypes.object),
	readOnly: PropTypes.bool.isRequired,
	update: PropTypes.func.isRequired,
	tag: PropTypes.func.isRequired
};
