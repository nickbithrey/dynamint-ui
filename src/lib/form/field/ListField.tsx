import * as React from 'react';
import { updateOnChange, ITypedField } from './Field';
import { IButton, ButtonContainer } from 'lib/button';
import { Key, Condition } from '~/lib';
import Table, { ITableColumn } from 'lib/table';
import { Selector, ISelectionDetails, ISelectionButton, SelectionType } from 'lib/selector';

export interface IListField extends ITypedField<Array<Object>> {
    storeKey: string;
    loading: boolean;
    entityForm: React.ComponentType;
    columns: Array<ITableColumn>;
}

const createButton: ISelectionButton & Condition<any> = {
    button: {
        key: 'create',
        content: 'Create',
        link: '/compconfigs',
        type: 'default'
    },
    condition: (input: ISelectionDetails) => input.selected.length === 0
};

interface IListFieldState {
    items: Array<any>;
    selectedItem?: any;
    selectedIndex?: number;
}

export class ListField extends React.Component<ITypedField<Array<Object>>, IListFieldState> {

    constructor(props: ITypedField<Array<Object>>) {
        super(props);
        this.state = {items: props.value};
        this.buttons = this.buttons.bind(this);
        this.onEditDetails = this.onEditDetails.bind(this);
        this.renderForm = this.renderForm.bind(this);
        this.updateField = this.updateField.bind(this)
        this.clearEditDetails = this.clearEditDetails.bind(this);
        this.createNewItem = this.createNewItem.bind(this);
    }
    
    buttons(): Array<ISelectionButton & Condition<any>> {
        return [
            {
                button: {
                    key: 'edit', 
                    content: 'Edit',
                    type: 'primary',
                },
                onClickResolver: this.onEditDetails,
                condition: (input: ISelectionDetails) => input.selected.length > 0,
             },
             {
                 button: {
                     key: 'create',
                     content: 'Create',
                     type: 'default',
                     onClick: this.createNewItem
                 },
                 condition: (input: ISelectionDetails) => input.selected.length === 0
             }
        ];
    }
    
    createNewItem(): void {
        this.onEditDetails({
            selectedCount: 1,
            selected: [{}],
            selectedIndices: [this.state.items.length]
        })();
    }
    
    updateField(item: any): void {
        const list = [...this.state.items];
        list[this.state.selectedIndex] = item;
        this.props.update(this.props.name, list);
        this.clearEditDetails();
        this.setState({
            items: list
        });
    }
    
    onEditDetails(input: ISelectionDetails): () => void {
        return () => {
            this.setState({selectedItem: input.selected[0], selectedIndex: input.selectedIndices[0]});
        }
    }
    
    clearEditDetails(): void {
        this.setState({
            selectedItem: undefined,
            selectedIndex: undefined
        });
    } 
    
    render() {
        const {
            loading,
            columns,
            entityForm,
            storeKey
        } = this.props;
        const {
            items
        } = this.state;
        return (
            <>
                <Selector
                    storeKey={storeKey}
                    loading={loading}
                    items={items}
                    columns={columns}
                    buttons={this.state.selectedItem ? [] : this.buttons()}
                    selectionType={SelectionType.single}
                />
                {
                    this.renderForm(entityForm)
                }
            </>
        );
    }
    
    renderForm(entityForm: React.ComponentType<any>) {
        const Form = entityForm;
        if (this.state.selectedItem) {
            return <Form 
                item={this.state.selectedItem}
                update={this.updateField}
                back={this.clearEditDetails}
            />;
        }
        return null;
    }
    
}

