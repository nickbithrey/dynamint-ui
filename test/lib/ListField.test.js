import React from 'react';
import { shallow } from 'enzyme';

import ListField, { listUpdate } from '~/lib/ListField';
import { Checkbox } from 'office-ui-fabric-react/lib/Checkbox';

const Tag = () => <p>Tag</p>;

describe('<ListField /> rendering', () => {
	
	const props = {
		name: "name",
		label: "Label",
		value: [
			{field1: 'val11', field2: 'val12'},
			{field1: 'val21', field2: 'val22'}
		],
		update: jest.fn(() => []),
		tag: Tag,
		columns: [
			{key: 'field1', name: 'Field 1', fieldName: 'field1'},
			{key: 'field2', name: 'Field 2', fieldName: 'field2'}
		],
		newElement: {field1: 'val31', field2: 'val32'}
	};
	
	let wrapper;
	
	beforeEach(() => {
		wrapper = shallow( <ListField {...props} /> );
	})

	it('matches snapshot', () => {
		expect(wrapper).toMatchSnapshot();
	});
	
	it('adds new element prop to value on Add button click and calls update prop function', () => {
		expect(props.update.mock.calls).toHaveLength(0);
		const event = {
			preventDefault: () => []
		}
		wrapper.find('Button').simulate('click', event);
		expect(props.update.mock.calls).toHaveLength(1);
		expect(props.update.mock.calls[0][0]).toEqual({name: props.name, value: [...props.value, props.newElement]});
	});
	
	it('doesn\'t render details of row when unselected', () => {
		const rowProps = {
			selection: wrapper.find('StyledWithViewportComponent').props().selection,
			itemIndex: 1
		};
		const fieldUpdate = jest.fn(() => []);
		expect(wrapper.instance().renderRow(fieldUpdate, Tag)(rowProps)).toMatchSnapshot();
	});
		
	it('renders details of row when selected', () => {
		const rowProps = {
			selection: wrapper.find('StyledWithViewportComponent').props().selection,
			itemIndex: 1
		};
		const fieldUpdate = jest.fn(() => []);
		
		// call selection changed event
		const selection = wrapper.find('StyledWithViewportComponent').props().selection;
		selection._items = props.value;
		selection.toggleIndexSelected(1);
		selection._onSelectionChanged();
		
		expect(wrapper.instance().renderRow(fieldUpdate, Tag)(rowProps)).toMatchSnapshot();
	});
	
	it('passes update prop to selected row', () => {
		const rowProps = {
			selection: wrapper.find('StyledWithViewportComponent').props().selection,
			itemIndex: 1
		};
		const fieldUpdate = jest.fn(() => []);
		
		// call selection changed event
		const selection = wrapper.find('StyledWithViewportComponent').props().selection;
		selection._items = props.value;
		selection.toggleIndexSelected(1);
		selection._onSelectionChanged();
		
		const rowWrapper = shallow(wrapper.instance().renderRow(fieldUpdate, Tag)(rowProps));
		expect(rowWrapper.find('ListUpdate').prop('fieldUpdate')).toBe(fieldUpdate);
	});
	
	it('renders correct column details', () => {
		const item = {field1: 'val11', field2: 'val12', field3: true};
		const textColumn = {key: 'field2', name: 'Field 2', fieldName: 'field2'};
		const boolColumn = {key: 'field3', name: 'Field 3', fieldName: 'field3', type: 'boolean'};

		expect(wrapper.instance().renderColumn(item, 1, textColumn)).toEqual(<span>val12</span>);
		expect(wrapper.instance().renderColumn(item, 1, boolColumn)).toEqual(<span><Checkbox checked={true} /></span>);
	});
	
	it('sets defaults correctly', () => {
		const propsLessDefaults = {...props};
		Object.keys(ListField.defaultProps).forEach(k => {
			delete propsLessDefaults[k];
		});
		wrapper = shallow( <ListField {...propsLessDefaults} /> );
		
		Object.entries(ListField.defaultProps).forEach(([key,value]) => {
			expect(wrapper.instance().props[key]).toBe(value);
		});
		
	});
	
});
	
describe('<ListUpdate /> rendering', () => {
	
	const props = {
		fieldUpdate: jest.fn(() => []),
		itemIndex: 1,
		selection: {}
	};
	
	let wrapper;
	
	beforeEach(() => {
		const Component = listUpdate(Tag);
		wrapper = shallow(<Component {...props} />);
	})

	it('matches snapshot', () => {
		expect(wrapper).toMatchSnapshot();
	});
		
	it('calls update prop function on submit button click', () => {
		expect(props.fieldUpdate.mock.calls).toHaveLength(0);
		const event = {
			preventDefault: () => []
		}
		const state = {key: 'val'};
		wrapper.setState(state);
		wrapper.find('Button').simulate('click', event);
		expect(props.fieldUpdate.mock.calls).toHaveLength(1);
		expect(props.fieldUpdate.mock.calls[0][0]).toEqual(state);
		expect(props.fieldUpdate.mock.calls[0][1]).toBe(props.itemIndex);
	});
	
});
	
