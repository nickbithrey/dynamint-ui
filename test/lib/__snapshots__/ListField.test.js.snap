// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ListField /> rendering doesn't render details of row when unselected 1`] = `
<StyledDetailsRowBase
  itemIndex={1}
  selection={
    Selection {
      "_anchoredIndex": 0,
      "_canSelectItem": [Function],
      "_changeEventSuppressionCount": 0,
      "_exemptedCount": 0,
      "_exemptedIndices": Object {},
      "_getKey": [Function],
      "_isModal": false,
      "_items": Array [],
      "_keyToIndexMap": Object {},
      "_onSelectionChanged": [Function],
      "_selectedIndices": Array [],
      "_selectedItems": null,
      "_unselectableCount": 0,
      "_unselectableIndices": Object {},
      "count": 0,
      "mode": 2,
    }
  }
/>
`;

exports[`<ListField /> rendering matches snapshot 1`] = `
<div>
  <Rowise
    tokens={
      Object {
        "childrenGap": 5,
      }
    }
  >
    <Block>
      <StyledWithViewportComponent
        ariaLabelForSelectAllCheckbox="Toggle selection for all items"
        ariaLabelForSelectionColumn="Toggle selection"
        columns={
          Array [
            Object {
              "fieldName": "field1",
              "key": "field1",
              "name": "Field 1",
            },
            Object {
              "fieldName": "field2",
              "key": "field2",
              "name": "Field 2",
            },
          ]
        }
        compact={false}
        enterModalSelectionOnTouch={true}
        isHeaderVisible={true}
        items={
          Array [
            Object {
              "field1": "val11",
              "field2": "val12",
            },
            Object {
              "field1": "val21",
              "field2": "val22",
            },
          ]
        }
        layoutMode={1}
        onRenderItemColumn={[Function]}
        onRenderRow={[Function]}
        selection={
          Selection {
            "_anchoredIndex": 0,
            "_canSelectItem": [Function],
            "_changeEventSuppressionCount": 0,
            "_exemptedCount": 0,
            "_exemptedIndices": Object {},
            "_getKey": [Function],
            "_isModal": false,
            "_items": Array [],
            "_keyToIndexMap": Object {},
            "_onSelectionChanged": [Function],
            "_selectedItems": null,
            "_unselectableCount": 0,
            "_unselectableIndices": Object {},
            "count": 0,
            "mode": 2,
          }
        }
        selectionMode={1}
        selectionPreservedOnEmptyClick={true}
        setKey="set"
      />
    </Block>
    <Block>
      <Button
        condition={[Function]}
        onClick={[Function]}
        text="Add"
        type="standard"
      />
    </Block>
  </Rowise>
</div>
`;

exports[`<ListField /> rendering renders details of row when selected 1`] = `
<div>
  <StyledDetailsRowBase
    itemIndex={1}
    selection={
      Selection {
        "_anchoredIndex": 1,
        "_canSelectItem": [Function],
        "_changeEventSuppressionCount": 0,
        "_exemptedCount": 1,
        "_exemptedIndices": Object {
          "1": true,
        },
        "_getKey": [Function],
        "_hasChanged": false,
        "_isModal": false,
        "_items": Array [
          Object {
            "field1": "val11",
            "field2": "val12",
          },
          Object {
            "field1": "val21",
            "field2": "val22",
          },
        ],
        "_keyToIndexMap": Object {},
        "_onSelectionChanged": [Function],
        "_selectedIndices": Array [
          1,
        ],
        "_selectedItems": null,
        "_unselectableCount": 0,
        "_unselectableIndices": Object {},
        "count": 1,
        "mode": 2,
      }
    }
  />
  <ListUpdate
    fieldUpdate={[MockFunction]}
    itemIndex={1}
    selection={
      Selection {
        "_anchoredIndex": 1,
        "_canSelectItem": [Function],
        "_changeEventSuppressionCount": 0,
        "_exemptedCount": 1,
        "_exemptedIndices": Object {
          "1": true,
        },
        "_getKey": [Function],
        "_hasChanged": false,
        "_isModal": false,
        "_items": Array [
          Object {
            "field1": "val11",
            "field2": "val12",
          },
          Object {
            "field1": "val21",
            "field2": "val22",
          },
        ],
        "_keyToIndexMap": Object {},
        "_onSelectionChanged": [Function],
        "_selectedIndices": Array [
          1,
        ],
        "_selectedItems": null,
        "_unselectableCount": 0,
        "_unselectableIndices": Object {},
        "count": 1,
        "mode": 2,
      }
    }
  />
</div>
`;

exports[`<ListUpdate /> rendering matches snapshot 1`] = `
<div>
  <Rowise
    tokens={
      Object {
        "childrenGap": 5,
      }
    }
  >
    <Block>
      <Tag
        updateField={[Function]}
      />
    </Block>
    <Block>
      <ButtonsContainer
        buttons={
          Array [
            Object {
              "condition": [Function],
              "onClick": [Function],
              "text": "Submit",
              "type": "primary",
            },
            Object {
              "condition": [Function],
              "onClick": [Function],
              "text": "Remove",
              "type": "standard",
            },
          ]
        }
      />
    </Block>
  </Rowise>
</div>
`;
